### PR TITLE
docs: ADR-010 enabling standalone ICS-02 integration

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
[📖 Rendered](https://github.com/cosmos/ibc-rs/blob/892c9ec35c90295a3ebdfc6b61f5287cd4bddb15/docs/architecture/adr-010-enable-standalone-ics02-integration.md)

Part of: #1114

Implementation PR: https://github.com/cosmos/ibc-rs/pull/1115

______

### PR author checklist:
- [X] Linked to GitHub issue.
- [X] Updated code comments and documentation (e.g., `docs/`).
- [X] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+Here is a high-level diagram that illustrates the optimal boundaries for each
+context trait, specifying the scope of access by each of these contexts. In
+addition, the classification of APIs should be such that they allow the
+independent integration of light client implementations.
+
+<div style="text-align: center;">
+  <img src="./assets/adr010.png" alt="Main components" width="800" height="600">
+</div>
+
+The primary `ValidationContext` and `ExecutionContext` traits will be
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
The primary `ValidationContext` and `ExecutionContext` traits at the ICS-24 host level will be
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+pub trait ValidationContext {
+    type V: ClientValidationContext;
+-    type E: ClientExecutionContext;
+-    type AnyConsensusState: ConsensusState;
+-    type AnyClientState: ClientState<Self::V, Self::E>;
+
+    /// Retrieve the context that implements all clients' `ValidationContext`.
+    fn get_client_validation_context(&self) -> &Self::V;
+
+     /// This method will be removed and instead, the `ClientStateDecoder` trait will be introduced.
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
     // This method will be removed and replaced by a `ClientStateDecoder` trait that will encapsulate the ability to decode a client state from an `Any`
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+    fn delete_update_meta(
+        &mut self,
+        client_id: ClientId,
+        height: Height,
+    ) -> Result<(), ContextError>;
+}
+```
+
+The introduction of the `ClientStateMut` associated type additional to the
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
The introduction of the `ClientStateMut` associated type in addition to the
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+Also, the introduction of `<ClientStateRef = Self::ClientStateMut>` is prompted
+by the need to address the characteristics of concrete `ClientState`
+definitions. For instance, in the case of ICS-07, such as `TmClientState`, the
+struct definition can't be split into two fragments, one for validation and the
+other for execution. Therefore, contexts implementing `ClientExecutionContext`
+must introduce a `ClientStateMut` type the same as `ClientStateRef`.
+
+With the mentioned classification, we can now streamline ICS-07 specific APIs,
+eliminating the requirement for implementing a redundant consensus_state()
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
eliminating the requirement for implementing a redundant `consensus_state()`
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+method. For the sake of simplification, we can remove the CommonContext trait
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
method. For the sake of simplification, we can remove the `CommonContext` trait
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+and consolidate everything under the TmValidationContext as follows:
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
and consolidate everything under the `TmValidationContext` as follows:
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    impl<T> ClientStateDecoder for T where T: TryFrom<Any, Error = ClientError> {}
+
+    ```
+
+- We will maintain the `client_counter()` and `increase_client_counter()` methods
+  within the main context traits. This stems from the fact that light clients do
+  not rely on the relative positions in their processes. Additionally, these
+  counters are globally tracked, and only IBC handlers invoke these methods for
+  setting or retrieving client identifiers.
+- The `host_consensus_state()` renamed to the `self_consensus_state()` for
```
Comment (seanchen1991) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]:
```suggestion
- The `host_consensus_state()` will be renamed to `self_consensus_state()` in order to maintain
```

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+      understand their interactions and roles. For instance some client relevant
+      methods are located under the client contexts, while others are under the
+      main (ICS-24) contexts.
+
+3. **Minimizing associated types’ declarations**
+    - Our current APIs pose a challenge by requiring users to introduce multiple
+      yet the same associated types like AnyClientState, AnyConsensusState, etc
+      across different layers. A more efficient solution involves introducing
+      such types in fewer places, providing better access to the traits across
+      different layers.
```
Comment (rnbguy) [inline docs/architecture/adr-010-enable-standalone-ics02-integration.md]: Can you please comment here on what features we need from rustc to improve this later?

- associated_type_bounds
- associated_type_defaults
- implied_bounds

**Review by rnbguy — APPROVED** [review]
LGTM :+1:

One comment about listing future solutions for the negative consequences.

File: docs/architecture/adr-010-enable-standalone-ics02-integration.md (hunk 8 above)
Comment (Farhad-Shabani) [reply docs/architecture/adr-010-enable-standalone-ics02-integration.md]: [034604a](https://github.com/cosmos/ibc-rs/pull/1116/commits/034604aca6663a554ea1704b01fbbd6dd899789f)

**Pull request merged**